### PR TITLE
Upgrade to scalajs-react v1.0.0-RC1

### PR DESCRIPTION
**Pull request by netzwerg**
This is a starting point, but needs careful review (will add comments to document my main struggles).

### CODE REVIEW

File: diode-react/src/main/scala/diode/react/ReactConnector.scala [hunk 1]
```diff
@@ -115,22 +116,24 @@ trait ReactConnector[M <: AnyRef] { circuit: Circuit[M] =>
       }
 
       private def changeHandler(cursor: ModelRO[S]): Unit = {
-        // modify state if we are mounted and state has actually changed
-        if (t.isMounted() && modelReader =!= t.accessDirect.state) {
-          t.accessDirect.setState(modelReader())
-        }
+        val isMounted = t.isMounted
```
Comment (netzwerg) [inline diode-react/src/main/scala/diode/react/ReactConnector.scala]: This is super awkward – I am confident that you will find a smoother way of composing these callbacks...

File: diode-react/src/main/scala/diode/react/ReactConnector.scala [hunk 2]
```diff
@@ -115,22 +116,24 @@ trait ReactConnector[M <: AnyRef] { circuit: Circuit[M] =>
...
       .renderBackend[Backend]
-      .componentWillMount(scope => scope.backend.willMount)
-      .componentWillUnmount(scope => scope.backend.willUnmount)
-      .shouldComponentUpdate(scope => (scope.currentState ne scope.nextState) || (scope.currentProps ne scope.nextProps))
+      .componentWillMount(_.backend.willMount)
+      .componentWillUnmount(_.backend.willUnmount)
+      .shouldComponentUpdatePure(scope => (scope.currentState ne scope.nextState) || (scope.currentProps ne scope.nextProps))
       .build
-      .set(key)
+      .withRawProp("key", key)
```
Comment (netzwerg) [inline diode-react/src/main/scala/diode/react/ReactConnector.scala]: The new API also offers `withKey(k: Key)`, so maybe you want to change the public API of `connect` to use a `react.Key`?

File: diode-react/src/main/scala/diode/react/package.scala [hunk 3]
```diff
@@ -1,7 +1,9 @@
 package diode
 
 import japgolly.scalajs.react._
+import japgolly.scalajs.react.component.Generic.Unmounted
+import japgolly.scalajs.react.vdom.VdomElement
 
 package object react {
-  type ReactConnectProxy[A] = ReactComponentC.ReqProps[ModelProxy[A] => ReactElement, A, _, TopNode]
+  type ReactConnectProxy[A] = CtorType.Props[ModelProxy[A] => VdomElement, Unmounted[ModelProxy[A] => VdomElement, _]]
```
Comment (netzwerg) [inline diode-react/src/main/scala/diode/react/package.scala]: This was my main struggle – are the new types correct?

**Comment by teozkr** [comment]
This should  should probably be an implicit `Unmounted[_, _] => VdomElement` defined, since that's what `myComponent(props)` now returns.

**Comment by netzwerg** [comment]
Isn't this already defined by 2nd type argument of `CtorType.Props`?

```scala
type ReactConnectProxy[A] = CtorType.Props[ModelProxy[A] => VdomElement, Unmounted[ModelProxy[A] => VdomElement, _]]
```

I tested the upgrade in a toy project and was able to `.renderIntoDOM` without needing to call `.vdomElement` explicitly, i.e. this compiled and worked as expected:

```scala
val modelConnection = new Circuit().connect(m => m)
val component = modelConnection(proxy =>
  <.div(
    ImageView(proxy.zoom(_.image)),
    DetailsView(proxy.zoom(_.details)),
    HistogramView(proxy.zoom(_.histogram))
  )
)
component.renderIntoDOM(document.getElementById("root"))
```

**Comment by teozkr** [comment]
Yeah, that works, but the following currently doesn't:

```scala
val modelConnection = new Circuit().connect(m => m.image)
val component = modelConnection(proxy =>
    ImageView(proxy)
)
component.renderIntoDOM(document.getElementById("root"))
```

**Comment by netzwerg** [comment]
This works for me.

> This should probably also define an implicit Unmounted[_, _] => VdomElement

See [Implicits.scala, Line 98](https://github.com/japgolly/scalajs-react/blob/master/core/src/main/scala/japgolly/scalajs/react/vdom/Implicits.scala#L98):

```scala
trait ImplicitsForVdomElement {
  implicit def vdomElementFromComponent(u: Generic.UnmountedWithRoot[_, _, _, _]): VdomElement =
    u.vdomElement
}
```

**Comment by teozkr** [comment]
That worked, whoops!

However, annoyingly, IDEA still seems to consider it a useless import and wants to remove it whenever I ask it to autoformat the code.

**Comment by netzwerg** [comment]
Sounds annoying. You might want to try the latest EAP:

```
IntelliJ IDEA 2017.1 EAP
Build #IU-171.3691.12, built on March 1, 2017
```

with Scala Plugin 2017.1.9 – works like a charm.

**Comment by teozkr** [comment]
Yup, everything is fine with the EAP. Thanks!

**Comment by ikamthania** [comment]
Any update on this pr??

**Comment by ochrons** [comment]
Are people using this successfully in their scalajs-react 1.0.0 apps? There's also RC2 out now :)

One thing is that many Diode users may not be upgrading to SJSReact 1.0.0 very quickly, so support for the old 0.11.x branch should be maintained.

I'll make a separate branch for this in the main repo, so that it can be developed further (ie. updating the documentation etc) without breaking the master. You can find it at https://github.com/suzaku-io/diode/tree/scalajs-react-1.0.0

@netzwerg can you retarget the PR to that branch?

**Comment by teozkr** [comment]
@ochrons I've been using it with 1.0.0-RC1 without issues (aside from the one I wrote about before).

**Comment by bpatters** [comment]
I'm using a private build of this + my PR with RC1 for my application, works great. For RC2 there is a minor change or two required for this PR to work. I think your idea of a new branch is a good idea until 1.0.0 is officially released for a short period.

**Comment by netzwerg** [comment]
Ok, I retargeted the PR. I won't have time to upgrade to RC2 this week, but as @bpatters pointed out, it shouldn't require many changes.

**Comment by ochrons** [comment]
Ok, I merged this, so others can also commit new PRs on top of it in the special branch.

**Pull request merged**